### PR TITLE
Reimplement setting source classes

### PR DESCRIPTION
**Pull request by hramezani**
Reimplement pydantic-settings based on the idea proposed in https://github.com/pydantic/pydantic-settings/pull/10#pullrequestreview-1222893531

- Added `PydanticBaseSource` abc
- Other source classes changed to inherit from abc
- Added `DotEnvSettingsSource` inherit from `EnvSettingsSource`

### CODE REVIEW

File: pydantic_settings/main.py [hunk 1]
```diff
@@ -134,98 +153,166 @@ def parse_env_var(cls, field_name: str, raw_val: str) -> Any:
     __config__: ClassVar[Type[Config]]
 
 
-class InitSettingsSource:
+class PydanticBaseSource(ABC):
```
Comment (samuelcolvin) [inline pydantic_settings/main.py:156]: please can you add a docstring here.

File: pydantic_settings/main.py [hunk 2]
```diff
@@ -134,98 +153,166 @@ def parse_env_var(cls, field_name: str, raw_val: str) -> Any:
...
+    def __init__(self, settings_cls: Type[BaseSettings]):
+        self.settings_cls = settings_cls
+        self.config = settings_cls.__config__
+        self.case_sensitive = self.config.case_sensitive
+
+    @abstractmethod
+    def get_field_value(self, field: ModelField) -> Any:
```
Comment (samuelcolvin) [inline pydantic_settings/main.py:163]: docstring, and methods below.

File: pydantic_settings/main.py (hunk 1 above)
Comment (samuelcolvin) [reply pydantic_settings/main.py:156]: should these all move to a separate `sources.py` module?

File: pydantic_settings/main.py [hunk 3]
```diff
@@ -134,98 +153,166 @@ def parse_env_var(cls, field_name: str, raw_val: str) -> Any:
...
+    def prepare_field(self, field_name: str, field: ModelField, value: Any) -> Any:
+        if field.is_complex():
+            return json.loads(value)
+        return value
+
+    def __call__(self) -> Dict[str, Any]:
+        d: Dict[str, Any] = {}
+
+        for name, field in self.settings_cls.__fields__.items():
+            field_value = self.get_field_value(field)
```
Comment (samuelcolvin) [inline pydantic_settings/main.py:175]: I think this too should be wrapped an exception.

File: pydantic_settings/main.py [hunk 4]
```diff
@@ -134,98 +153,166 @@ def parse_env_var(cls, field_name: str, raw_val: str) -> Any:
...
+
+            try:
+                field_value = self.prepare_field(name, field, field_value)
+            except ValueError as e:
+                raise SettingsError(f'error parsing value for field "{name}"') from e
```
Comment (samuelcolvin) [inline pydantic_settings/main.py:180]: this should probably say which source we're getting data from to to ease debugging.

File: pydantic_settings/main.py [hunk 5]
```diff
@@ -134,98 +153,166 @@ def parse_env_var(cls, field_name: str, raw_val: str) -> Any:
...
+    def __init__(self, settings_cls: Type[BaseSettings]):
+        self.settings_cls = settings_cls
+        self.config = settings_cls.__config__
+        self.case_sensitive = self.config.case_sensitive
+
+    @abstractmethod
+    def get_field_value(self, field: ModelField) -> Any:
+        pass
+
+    def prepare_field(self, field_name: str, field: ModelField, value: Any) -> Any:
```
Comment (samuelcolvin) [inline pydantic_settings/main.py:166]:
```suggestion
    def prepare_field_value(self, field_name: str, field: ModelField, value: Any) -> Any:
```

File: pydantic_settings/main.py [hunk 6]
```diff
@@ -134,98 +153,166 @@ def parse_env_var(cls, field_name: str, raw_val: str) -> Any:
...
         return self.init_kwargs
 
     def __repr__(self) -> str:
         return f'InitSettingsSource(init_kwargs={self.init_kwargs!r})'
 
 
-class EnvSettingsSource:
-    __slots__ = ('env_file', 'env_file_encoding', 'env_nested_delimiter', 'env_prefix_len')
+class SecretsSettingsSource(PydanticBaseSource):
+    __slots__ = ('secrets_dir',)
```
Comment (samuelcolvin) [inline pydantic_settings/main.py:206]:
```suggestion
```

No need to mess around with slots on this one.

File: pydantic_settings/main.py [hunk 7]
```diff
@@ -120,11 +137,13 @@ def prepare_field(cls, field: ModelField) -> None:
         @classmethod
         def customise_sources(
             cls,
-            init_settings: SettingsSourceCallable,
-            env_settings: SettingsSourceCallable,
-            file_secret_settings: SettingsSourceCallable,
-        ) -> Tuple[SettingsSourceCallable, ...]:
-            return init_settings, env_settings, file_secret_settings
+            settings_cls: Type['BaseSettings'],
```
Comment (samuelcolvin) [inline pydantic_settings/main.py:140]: either now or in a separate PR, let's add `from __future__ import annotations as _annotations` so these don't need to be strings.

**Review by samuelcolvin — COMMENTED** [review]
otherwise LGTM.

File: pydantic_settings/main.py (hunk 1 above)
Comment (hramezani) [reply pydantic_settings/main.py]: move to `sources.py` and docstrings added

File: pydantic_settings/main.py (hunk 3 above)
Comment (hramezani) [reply pydantic_settings/main.py]: wrapped and a test also added

File: pydantic_settings/main.py (hunk 4 above)
Comment (hramezani) [reply pydantic_settings/main.py]: Setting source added to error message and tests updated

File: pydantic_settings/main.py (hunk 7 above)
Comment (hramezani) [reply pydantic_settings/main.py]: Done

File: pydantic_settings/main.py [hunk 8]
```diff
@@ -1,23 +1,25 @@
...
 
 from pydantic.config import BaseConfig, Extra
 from pydantic.fields import ModelField
 from pydantic.main import BaseModel
-from pydantic.typing import StrPath, display_as_type, get_origin, is_union
-from pydantic.utils import deep_update, path_type, sequence_like
+from pydantic.typing import StrPath, display_as_type
+from pydantic.utils import deep_update, sequence_like
+
+from pydantic_settings.sources import (
```
Comment (samuelcolvin) [inline pydantic_settings/main.py:12]:
```suggestion
from .sources import (
```

**Review by samuelcolvin — APPROVED** [review]
one tiny thing, otherwise LGTM.

Once that's fixed, please merge.

**Pull request merged**